io/src/read_geometry.cpp

```diff
@@ -11,16 +11,74 @@
 #include "traccc/io/details/read_surfaces.hpp"
 #include "traccc/io/utils.hpp"
 
+// Detray include(s).
+#include <detray/core/detector.hpp>
+#include <detray/io/frontend/detector_reader.hpp>
+
+// VecMem include(s).
+#include <vecmem/memory/host_memory_resource.hpp>
+
+// System include(s).
+#include <stdexcept>
+
+namespace {
+
+/// Helper function constructing @c traccc::geometry from Detray JSON
+std::pair<traccc::geometry,
+          std::unique_ptr<std::map<std::uint64_t, detray::geometry::barcode>>>
+read_json_geometry(std::string_view filename) {
+
+    // Memory resource used while reading the detector JSON.
+    vecmem::host_memory_resource host_mr;
+
+    // The result objects.
+    traccc::geometry surface_transforms;
+    std::unique_ptr<std::map<std::uint64_t, detray::geometry::barcode>>
+        barcode_map;
+
+    {
+        // Construct a detector object.
+        detray::io::detector_reader_config reader_cfg{};
+        reader_cfg.add_file(std::string{filename});
+        auto [detector, _] =
+            detray::io::read_detector<detray::detector<>>(host_mr, reader_cfg);
+
+        // Construct an "old style geometry" from the detector object.
+        surface_transforms = traccc::io::alt_read_geometry(detector);
+
+        // Construct a map from Acts surface identifiers to Detray barcodes.
+        barcode_map = std::make_unique<
+            std::map<std::uint64_t, detray::geometry::barcode>>();
+        for (const auto& surface : detector.surfaces()) {
+            (*barcode_map)[surface.source] = surface.barcode();
+        }
+    }
+
+    // Return the created objects.
+    return {surface_transforms, std::move(barcode_map)};
+}
+
+}  // namespace
+
 namespace traccc::io {
 
-geometry read_geometry(std::string_view filename, data_format format) {
+std::pair<geometry,
+          std::unique_ptr<std::map<std::uint64_t, detray::geometry::barcode>>>
+read_geometry(std::string_view filename, data_format format) {
 
     // Construct the full file name.
     const std::string full_filename = data_directory() + filename.data();
 
-    // Read the file using another function. Relying on the auto-conversion of
-    // the output type of that other function.
-    return details::read_surfaces(full_filename, format);
+    // Decide how to read the file.
+    switch (format) {
```

Comment:
Can't we use the "detray switch" here directly?

Reply:
"Detray switch"? 😕 We have this `traccc::data_format` enum that is used in a number of different ways. But it did seem appropriate here. Where I decided that "CSV format" would mean the old style geometry reading (from a CSV file), and "JSON format" would mean Detray's file I/O. From an actual JSON file.

Or does Detray have functionality resembling that of [traccc::io::csv::read_surfaces(...)](https://github.com/acts-project/traccc/blob/main/io/src/csv/read_surfaces.cpp)? I don't think it does. 🤔

Reply:
I meant ```det_opts.use_detray_detector```, but its fine this way. 

No, in detray it would not make a lot of sense to just read in the transforms, as we need full coordinate transformations for different shapes